dev/TEAL.md

```diff
@@ -0,0 +1,219 @@
+| `*` | A times B. Panic on overflow. |
+| `<` | A less than B => {0 or 1} |
+| `>` | A greater than B => {0 or 1} |
+| `<=` | A less than or equal to B => {0 or 1} |
+| `>=` | A greater than or equal to B => {0 or 1} |
+| `&&` | A is not zero and B is not zero => {0 or 1} |
+| `\|\|` | A is not zero or B is not zero => {0 or 1} |
+| `==` | A is equal to B => {0 or 1} |
+| `!=` | A is not equal to B => {0 or 1} |
+| `!` | X == 0 yields 1; else 0 |
+| `len` | yields length of byte value |
+| `btoi` | converts bytes as big endian to uint64 |
+| `%` | A modulo B. Panic if B == 0. |
+| `\|` | A bitwise-or B |
+| `&` | A bitwise-and B |
+| `^` | A bitwise-xor B |
+| `~` | bitwise invert value |
+
+### Loading Values
+
+Opcodes for getting data onto the stack.
+
+Some of these have immediate data in the byte or bytes after the opcode.
+
+| Op | Description |
+| --- | --- |
+| `intcblock` | load block of uint64 constants |
+| `intc` | push value from uint64 constants to stack by index into constants |
+| `intc_0` | push uint64 constant 0 to stack |
+| `intc_1` | push uint64 constant 1 to stack |
+| `intc_2` | push uint64 constant 2 to stack |
+| `intc_3` | push uint64 constant 3 to stack |
+| `bytecblock` | load block of byte-array constants |
+| `bytec` | push bytes constant to stack by index into constants |
+| `bytec_0` | push bytes constant 0 to stack |
+| `bytec_1` | push bytes constant 1 to stack |
+| `bytec_2` | push bytes constant 2 to stack |
+| `bytec_3` | push bytes constant 3 to stack |
+| `arg` | push LogicSig.Args[N] value to stack by index |
+| `arg_0` | push LogicSig.Args[0] to stack |
+| `arg_1` | push LogicSig.Args[1] to stack |
+| `arg_2` | push LogicSig.Args[2] to stack |
+| `arg_3` | push LogicSig.Args[3] to stack |
+| `txn` | push field from current transaction to stack |
+| `gtxn` | push field to the stack from a transaction in the current transaction group |
+| `global` | push value from globals to stack |
+| `load` | copy a value from scratch space to the stack |
+| `store` | pop a value from the stack and store to scratch space |
+
+**Transaction Fields**
+
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | Sender | []byte |
+| 1 | Fee | uint64 |
+| 2 | FirstValid | uint64 |
+| 3 | LastValid | uint64 |
+| 4 | Note | []byte |
+| 5 | Receiver | []byte |
+| 6 | Amount | uint64 |
+| 7 | CloseRemainderTo | []byte |
+| 8 | VotePK | []byte |
+| 9 | SelectionPK | []byte |
+| 10 | VoteFirst | uint64 |
+| 11 | VoteLast | uint64 |
+| 12 | VoteKeyDilution | uint64 |
+| 13 | Type | []byte |
+| 14 | TypeEnum | uint64 |
+| 15 | XferAsset | []byte |
+| 16 | AssetAmount | uint64 |
+| 17 | AssetSender | []byte |
+| 18 | AssetReceiver | []byte |
+| 19 | AssetCloseTo | []byte |
+| 20 | GroupIndex | uint64 |
+
+
+**Global Fields**
+
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | Round | uint64 |
+| 1 | MinTxnFee | uint64 |
+| 2 | MinBalance | uint64 |
+| 3 | MaxTxnLife | uint64 |
+| 4 | TimeStamp | uint64 |
```

Comment:
Is this the current block's timestamp or the previous block's timestamp?

Reply:
current. I'm pretty sure I traced the execution and we set the new block-header TimeStamp at the beginning of running the evaluator so that it's available to scripts.

Reply:
Gotcha. Fine either way, although we might see some weirdness with the transaction pool because the block evaluator it uses for deciding whether to add a transaction to the pool doesn't know the upcoming block's timestamp yet. (On the other hand, the pool buffers more than one block's worth of transactions, so under congestion it doesn't necessarily know the previous block's timestamp either.)